datadog/dogstatsd/base.py

```diff
@@ -362,12 +374,25 @@ def open_buffer(self, max_buffer_size=None):
     def close_buffer(self):
         """
         Flush the buffer and switch back to single metric packets.
+
+        Note: This method must be called after a matching open_buffer()
+        invocation.
         """
-        self._send = self._send_to_server
 
-        if self.buffer:
-            # Only send packets if there are packets to send
-            self._flush_buffer()
+        try:
+            if not hasattr(self, 'buffer'):
+                raise BufferError('Cannot close buffer that was never opened')
+
+            self._send = self._send_to_server
+
+            if self.buffer:
+                # Only send packets if there are packets to send
+                self._flush_buffer()
+        finally:
+            # Release all locks that might have been held by our thread.
+            while self._buffer_lock_depth > 0:
+                self._buffer_lock_depth -= 1
+                self._buffer_lock.release()
```

Comment:
Does this mean you can have multiple `open_buffer` from the same thread and only one `close_buffer`?
That sounds like trying to circumvent programming errors.

Reply:
> Does this mean you can have multiple open_buffer from the same thread and only one close_buffer?
That sounds like trying to circumvent programming errors.

There was no guidance that an `open_buffer` should match a `close_buffer` should match a `close_buffer`. Even if we did have that guidance, it's not technically a programming error - the following code gets into the tricky territory of "what's the correct behavior expected" that ends up being ambiguous as to what it should do/allow:
```python
statds.open_buffer()
# do something to the buffer

statsd.open_buffer()
# do something else
statsd.close_buffer()
```
The second `open_buffer` in the strictest of terms (which is what you were pointing out I think) seems like it should just throw an exception as it's out-of-order but given the former behavior and its user friendliness it could be interpreted as a "just give me a new buffer to write". We've discussed this a bit on slack before the PR was implemented and chose the latter as a more-appropriate behavior.

As a sidenote, both `open_buffer` and `close_buffer` are getting deprecated soon since we want buffering by default so they will effectively be `noop`s in the future.